repo.go

```diff
@@ -37,12 +37,138 @@ type Interface interface {
 // Charm store references will use the provided parameters.
 // Local references will use the provided path.
 func InferRepository(ref *charm.URL, charmStoreParams NewCharmStoreParams, localRepoPath string) (Interface, error) {
-	switch ref.Schema {
+	dis := NewDispatcher(charmStoreParams, localRepoPath)
+	inferred, err := dis.Infer(ref.Schema)
+	if err != nil {
+		return nil, err
+	}
+	repo, err := inferred.NewRepo()
+	if err != nil {
+		return nil, err
+	}
+	return repo, nil
+}
+
+// Dispatcher supports selecting and performing repo-related behavior
+// based on charm schemas. Behavior selection is independent of the
+// source of the schema (e.g. charm.URL).
+//
+// Recognized schemas are:
+//   cs
+//   local
+type Dispatcher struct {
+	// Factory is the charm repo factory that the dispatcher uses.
+	Factory Factory
+
+	// Handlers maps charm schemas to functions that will be called
+	// in response to the corresponding schema. A missing entry
+	// (schema not found in the map) indicates that the schema is not
+	// supported. If the entry is there but holds a nil function
+	// then that indicates a no-op.
+	Handlers map[string]func(schema string, repo Interface) error
+}
+
+// NewDispatcher returns a new repo dispatcher with no-op handlers for
+// each of the recognized schemas. The dispatcher's repo factory is
+// created using the provided charm store and local repo info.
+func NewDispatcher(charmStoreParams NewCharmStoreParams, localRepoPath string) *Dispatcher {
+	return &Dispatcher{
+		Factory:  NewFactory(charmStoreParams, localRepoPath),
+		Handlers: map[string]func(string, Interface) error{"cs": nil, "local": nil},
+	}
+}
+
+// Infer determines the repo and handler to use for the given schema.
+// It does not execute any behavior relative to the repo.
+func (dis Dispatcher) Infer(schema string) (*Inferred, error) {
```

Comment:
I don't understand the motivation for all this code. We're replacing a totally trivial 8 line function with 135 lines of new abstractions without any obvious reason.

From the PR description:

"This is especially significant where we want to infer the repo but provide an existing cs client rather than the params."

If you want to do that (I don't really understand why you'd want to), then InferRepository is not the right function to use, and building more abstractions into it won't make it right.

ISTM that these abstractions are all about pushing test-related code down to a place where it doesn't belong. We still have exactly two schemas; the InferRepository code is correct as it is currently. If we wanted to provide more, we could easily provide a much simpler abstraction. For example:

	
	// Schemas represents a set of charm URL schemas
	// keyed by schema name.
	// Each schema can be used to create a charmrepo Interface
	// by calling the function associated with the name.
	type Schemas map[string]func() (Interface, error)
	
	// DefaultSchemas returns the standard set of schemas.
	// Charm store references will use the provided charm store parameters.
	// Local references will use the provided path.
	func DefaultSchemas(charmStoreParams NewCharmStoreParams, localRepoPath string) Schemas {
		return Schemas{
			"cs": func() (Interface, error) {
				return NewCharmStore(p.CS), nil
			},
			"local": func() (Interface, error) {
				return NewLocalRepository(localRepoPath)
			},
		}
	}
	
	// InferRepository returns a charm repository inferred from the provided charm
	// or bundle reference using the schemas found in s.
	func (s Schemas) InferRepository(ref *charm.URL) (Interface, error) {
		if f := s[ref.Schema]; f != nil {
			return f()
		}
		return nil, fmt.Errorf("unknown schema for charm reference %q", ref)
	}